core/src/main/java/org/apache/accumulo/core/iterators/system/LocalityGroupIterator.java

```diff
@@ -97,75 +134,122 @@ public static final int seek(HeapIterator hiter, LocalityGroup[] groups, Set<Byt
         } else {
-          // must include, if all excluded column families are in other locality groups
-          // then there are not unwanted column families in default LG
-          include = true;
+          // must include the default group as it may include cfs not in our cfSet
+          groupsToUse.add(groups.defaultGroup);
+        }
+      }
+
+      /*
+       * Need to consider the following cases for inclusive and exclusive (lgcf:locality group column family set, cf:column family set) lgcf and cf are disjoint
+       * lgcf and cf are the same cf contains lgcf lgcf contains cf lgccf and cf intersect but neither is a subset of the other
+       */
+      if (!inclusive) {
+        for (Entry<ByteSequence,LocalityGroup> entry : groups.groupByCf.entrySet()) {
+          if (!cfSet.contains(entry.getKey())) {
+            groupsToUse.add(entry.getValue());
+          }
+        }
+      } else if (groups.groupByCf.size() <= cfSet.size()) {
+        for (Entry<ByteSequence,LocalityGroup> entry : groups.groupByCf.entrySet()) {
+          if (cfSet.contains(entry.getKey())) {
+            groupsToUse.add(entry.getValue());
+          }
         }
       } else {
-        /*
-         * Need to consider the following cases for inclusive and exclusive (lgcf:locality group column family set, cf:column family set) lgcf and cf are
-         * disjoint lgcf and cf are the same cf contains lgcf lgcf contains cf lgccf and cf intersect but neither is a subset of the other
-         */
-
-        for (Entry<ByteSequence,MutableLong> entry : lgr.columnFamilies.entrySet())
-          if (entry.getValue().longValue() > 0)
-            if (cfSet.contains(entry.getKey())) {
-              if (inclusive)
-                include = true;
-            } else if (!inclusive) {
-              include = true;
-            }
+        for (ByteSequence cf : cfSet) {
+          LocalityGroup group = groups.groupByCf.get(cf);
+          if (group != null) {
+            groupsToUse.add(group);
+          }
+        }
       }
+    }
 
-      if (include) {
-        lgr.getIterator().seek(range, EMPTY_CF_SET, false);
-        hiter.addSource(lgr.getIterator());
-        numLGSeeked++;
-      }// every column family is excluded, zero count, or not present
+    for (LocalityGroup lgr : groupsToUse) {
+      lgr.getIterator().seek(range, EMPTY_CF_SET, false);
+      hiter.addSource(lgr.getIterator());
+      numLGSeeked++;
+    }
+
+    if (used != null) {
+      used.addAll(groupsToUse);
     }
 
     return numLGSeeked;
   }
 
   @Override
   public void seek(Range range, Collection<ByteSequence> columnFamilies, boolean inclusive) throws IOException {
-    seek(this, groups, nonDefaultColumnFamilies, range, columnFamilies, inclusive);
+
+    // determine if the arguments have changed since the last time
+    boolean sameArgs = false;
+    ImmutableSet<ByteSequence> cfSet = null;
+    if (lastUsed != null && inclusive == lastInclusive) {
```

Comment:
@ivakegg  my memory is a bit fuzzy on this.  Its good that InMemoryMapTest covers these changes.  However, I think some of these optimizations you made will benefit iterators seeking the in mem map, but not iterators seeking RFiles.  I think I was not surprised at lack of coverage, but at the lack of use of your optimizations by RFile.  Personally I would like to see RFile refactored to use your optimizations.